internal/util/ctxutil/ctxutil.go

```diff
@@ -61,20 +62,28 @@ func SleepWithJitter(ctx context.Context, d time.Duration, attempts int64) {
 	<-sleepCtx.Done()
 }
 
-// DurationWithJitter returns an exponential backoff duration based on retry with random jitter.
-// The maximum sleep is the cap. The minimum duration is at least 100 milliseconds.
+// DurationWithJitter returns an exponential backoff duration based on retry with random "full jitter".
+// https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
 //
-// Math/rand is good enough because we don't need the randomness to be cryptographically secure.
+// The maximum sleep is the cap. The minimum sleep is at least 3 milliseconds.
+// Provided cap must be larger than minimum sleep, and retry number must be a positive number.
 func DurationWithJitter(cap time.Duration, retry int64) time.Duration {
 	const base = time.Millisecond * 100
+	const minMilliseconds = 3
 
 	if retry < 1 {
 		panic("retry must be nonzero positive number")
 	}
 
+	if cap.Milliseconds() <= minMilliseconds {
+		panic(fmt.Sprintf("cap must be larger than min sleep (%dms)", minMilliseconds))
+	}
+
 	maxMilliseconds := float64(base.Milliseconds()) * math.Pow(2, float64(retry))
 	capMilliseconds := float64(cap.Milliseconds())
-	lowestValue := int64(math.Min(capMilliseconds, maxMilliseconds))
 
-	return time.Duration(rand.Int63n(lowestValue)) * time.Millisecond
+	lowestValue := int64(math.Min(capMilliseconds, maxMilliseconds)) - minMilliseconds
```

Comment:
I think I am a bit lost with calculation :)

As far as I understand, we use this one:
```
temp = min(cap, base * 2**attempt)
sleep = temp/2 + random_between(0, temp/2)
```

so, in our case `temp = min(capMilliseconds, maxMilliseconds)`
then what is the rest of the calculation about?

Reply:
The calculation you mentioned is the Equal Jitter

We use Full Jitter approach:
```
sleep = random_between(0, min(cap, base * 2 ** attempt))
```